### PR TITLE
Associate

### PR DESCRIPTION
**Pull request by NalianAchei2020**
> In this PR, I set up an association with the following requirements 📑
> 
> General requirements
 ✅ Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config). 
✅ Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
 ✅ Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> 
> Ruby requirements 
✅ Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).
> 
> Project requirements Create a class Classroom with the following: 
✅ @Label instance variable, should be initialized in the constructor. 
✅ Setter and getter for @Label (remember about attr_accessor). Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented: 
✅ Create the has-many side (a classroom has many students). 
✅ Create the belongs-to side (a student belongs to a classroom). 
✅ Make sure that when adding a student to a classroom it also sets the classroom for the student. 
✅ Make sure that when setting the classroom for a student it also adds it to the classrooms' students. Create a class Book with the following: 
✅ @title and https://github.com/author instance variables, should be initialized in the constructor. 
✅ Setters and getters for instance variables (remember about attr_accessor). Create a class Rental with the following: 
✅ @Date instance variable, should be initialized in the constructor.
 ✅ Setter and getter for @Date (remember about attr_accessor). Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
 ✅ Create the has-many side of Book and Rental (a book has many rentals).
 ✅ Create the belongs-to side of Rental and Book (a rental belongs to a book). 
✅ Create the has-many side of Person and Rental (a person has many rentals). 
✅ Create the belongs-to side of Rental and Person (a rental belongs to a person). 
✅ Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+require_relative 'trimmer_decorator'
+
+# Person class provides the default implementation of the correct_name method
+class Person < Nameable
+  def initialize(age, name = 'unknown', parent_permission: true)
+    super()
+    @id = Random.rand(1..1000)
+    @name = name
+    @age = age
+    @parent_permission = parent_permission
```
Comment (tufoinnkuo10) [inline person.rb:12]: - [ ] great job. please ensure that @rentals should be initialized as well

File: person.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  # Getter and setter in one
```
Comment (tufoinnkuo10) [inline person.rb]: - [ ] nice job. please kindy ensure
Getters and Setters for @rentals should be present
 ```ruby
  attr_accessor :name, :age, :rentals
```

File: person.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  attr_accessor :name, :age
+
+  # Getter
+  attr_reader :id
+
```
Comment (tufoinnkuo10) [inline person.rb:17]: - [ ] please note that A person has many rentals. Kindly ensure that A method for adding rentals should be present. here, the person itself should be sent as a parameter to create the new rental, achieving the 'has-many' association`
```ruby
def add_rental(book, date)
    Rental.new(date, book, self)
  end
```

File: students.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require_relative 'person'
+
```
Comment (tufoinnkuo10) [inline students.rb:2]: - [ ] please kindly ensure that When setting the classroom for a student it also adds it to the classrooms' students
`attr_reader :classroom`
- [ ] kindly add a method that serves as the 'belongs-to' side (a student belongs to a classroom)
```ruby
def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @NalianAchei2020 

There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight
✅ You used the correct Gitflow
✅ Descriptive pull request message
✅ linter error checks validated
✅ Documented your work professionally
✅ associations well set

Required Changes ♻️
Check the comments under the review
Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

**Review by Efiamotu-1 — APPROVED** [review]
Hi @NalianAchei2020, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# STATUS: APPROVED 🕺🥳 

image

## Highlights :dart: 
* Correct Gitflow :heavy_check_mark: 
* Good commit messages :heavy_check_mark: 
* All linters passing :heavy_check_mark: 
* Good PR summary :heavy_check_mark: 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, you can now Merge since it is Approved ✅, However if there are any conflicts please resolve them before merging.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request closed without merging**